pkg/whep/session.go

```diff
@@ -0,0 +1,387 @@
+package whep
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"io"
+	"strings"
+	"sync"
+
+	google_protobuf "google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/pion/rtcp"
+	"github.com/pion/rtp"
+	"github.com/pion/webrtc/v3"
+
+	"github.com/livekit/protocol/logger"
+	"github.com/livekit/protocol/rpc"
+	sdputil "github.com/livekit/protocol/sdp"
+	"github.com/livekit/protocol/tracer"
+	"github.com/livekit/protocol/utils"
+	"github.com/livekit/psrpc"
+
+	lksdk "github.com/livekit/server-sdk-go/v2"
+)
+
+var (
+	ErrResourceNotFound = psrpc.NewErrorf(psrpc.NotFound, "WHEP resource not found")
+)
+
+type SessionParams struct {
+	Logger      logger.Logger
+	Token       string
+	Participant string
+	Offer       string
+	WsUrl       string
+	RtcAPI      *webrtc.API
+}
+
+type Session struct {
+	params SessionParams
+
+	room              *lksdk.Room
+	pc                *webrtc.PeerConnection
+	lock              sync.Mutex
+	forwardingSenders map[string]*webrtc.RTPSender // RemoteTrack.SID -> Local RTPSender
+	availableSenders  []*webrtc.RTPSender
+	closed            bool
+
+	onClose func()
+}
+
+func NewSession(params SessionParams) *Session {
+	return &Session{
+		params:            params,
+		forwardingSenders: make(map[string]*webrtc.RTPSender),
+	}
+}
+
+func (s *Session) CreateAnswer() (string, error) {
+	pc, err := s.params.RtcAPI.NewPeerConnection(webrtc.Configuration{SDPSemantics: webrtc.SDPSemanticsUnifiedPlan})
+	if err != nil {
+		return "", err
+	}
+	s.pc = pc
+
+	offerSdp := webrtc.SessionDescription{
+		Type: webrtc.SDPTypeOffer,
+		SDP:  s.params.Offer,
+	}
+
+	// create track placeholders for forwarding sender
+	parsedOffer, err := offerSdp.Unmarshal()
+	if err != nil {
+		return "", err
+	}
+	for _, media := range parsedOffer.MediaDescriptions {
+		switch media.MediaName.Media {
+		case "audio":
+			track, err := webrtc.NewTrackLocalStaticRTP(webrtc.RTPCodecCapability{
+				MimeType:    webrtc.MimeTypeOpus,
+				ClockRate:   48000,
+				Channels:    2,
+				SDPFmtpLine: "minptime=10;useinbandfec=1",
+			}, utils.NewGuid(utils.TrackPrefix), s.params.Participant)
+			if err != nil {
+				return "", err
+			}
+			if t, err := pc.AddTransceiverFromTrack(track, webrtc.RTPTransceiverInit{Direction: webrtc.RTPTransceiverDirectionSendonly}); err != nil {
+				return "", err
+			} else {
+				s.availableSenders = append(s.availableSenders, t.Sender())
+			}
+
+		case "video":
+			track, err := webrtc.NewTrackLocalStaticRTP(webrtc.RTPCodecCapability{
+				MimeType:  webrtc.MimeTypeVP8,
```

Comment:
it is just a placeholder here to let pion create a sendonly transceiver, we'll replace it with desired track when remote track is subscribed

Reply:
Got it. Was wondering about SDP. We do `CreateAnswer` after adding these. That would advertise VP8 first I guess. If H.264 is used later, how would the SDP look with replace track?

Reply:
there is no renegotiation in whip/whep protocols so we need to negotiate all transceivers in first request. we list all supported codecs in the answer and sending actual payload when track is subscribed, it is ok since the first one (VP8) only means preferred codec but sends latter one is legal and client should prepare to receive it.

Reply:
Got it. Good to know that client is handling it.